route53/client.go

```diff
@@ -0,0 +1,248 @@
+package route53
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	r53 "github.com/aws/aws-sdk-go/service/route53"
+	"github.com/libdns/libdns"
+)
+
+// NewSession initializes the AWS client
+func (p *Provider) NewSession() error {
+	if p.MaxRetries == 0 {
+		p.MaxRetries = 5
+	}
+	sess, err := session.NewSession(&aws.Config{
+		MaxRetries: aws.Int(p.MaxRetries),
+	})
+	if err != nil {
+		return err
+	}
+
+	p.client = r53.New(sess)
+
+	return nil
+}
+
+func (p *Provider) getRecords(ctx context.Context, zoneID string) ([]libdns.Record, error) {
+	getRecordsInput := &r53.ListResourceRecordSetsInput{
+		HostedZoneId: aws.String(zoneID),
+		MaxItems:     aws.String("1000"),
+	}
+
+	var records []libdns.Record
+	var recordSets []*r53.ResourceRecordSet
+
+	for {
+		getRecordResult, err := p.client.ListResourceRecordSetsWithContext(ctx, getRecordsInput)
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				switch aerr.Code() {
+				case r53.ErrCodeNoSuchHostedZone:
+					return records, fmt.Errorf("%s: %s", r53.ErrCodeNoSuchHostedZone, aerr.Error())
+				case r53.ErrCodeInvalidInput:
+					return records, fmt.Errorf("%s: %s", r53.ErrCodeInvalidInput, aerr.Error())
+				default:
+					return records, fmt.Errorf(aerr.Error())
+				}
+			} else {
+				return records, fmt.Errorf(err.Error())
+			}
+		}
+
+		recordSets = append(recordSets, getRecordResult.ResourceRecordSets...)
+		if *getRecordResult.IsTruncated {
+			getRecordsInput.StartRecordName = getRecordResult.NextRecordName
+			getRecordsInput.StartRecordType = getRecordResult.NextRecordType
+			getRecordsInput.StartRecordIdentifier = getRecordResult.NextRecordIdentifier
+		} else {
+			break
+		}
+	}
+
+	for _, rrset := range recordSets {
+		for _, rrsetRecord := range rrset.ResourceRecords {
+			record := libdns.Record{
+				Name:  *rrset.Name,
+				Value: *rrsetRecord.Value,
+				Type:  *rrset.Type,
+				TTL:   time.Duration(*rrset.TTL) * time.Second,
+			}
```

Comment:
Is there any sort of "ID" associated with a record, from the Route53-side? i.e. does Route53 assign its own IDs to records? If so, it should be set in this struct too.

Reply:
Yeah, I looked at that and they don't have any IDs for recordSets as per https://docs.aws.amazon.com/sdk-for-go/api/service/route53/#ResourceRecordSet

The way it works is you need to submit a [changeBatch](https://docs.aws.amazon.com/sdk-for-go/api/service/route53/#ChangeBatch) and in this struct you define all the the recordSet info... so for example, to DELETE a record, you submit a `changeBatch` with the `Name`, `Value`, `TTL` and `Type` of the recordSet and they do their sourcery under the hood.

Reply:
Ah, interesting -- no problem then.